pkg/pipeline/conntrack/hash.go

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2022 IBM, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package contrack
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"hash/fnv"
+
+	"github.com/netobserv/flowlogs-pipeline/pkg/api"
+	"github.com/netobserv/flowlogs-pipeline/pkg/config"
+)
+
+// ComputeHash computes the hash of a flow log according to keyFields.
+// Two flow logs will have the same hash if they belong to the same connection.
+func ComputeHash(flowLog config.GenericMap, keyFields api.KeyFields) ([]byte, error) {
+	type hashType []byte
+	fieldGroup2hash := make(map[string]hashType)
+
+	// Compute the hash of each field group
+	for _, fg := range keyFields.FieldGroups {
+		h, err := computeHashFields(flowLog, fg.Fields)
+		if err != nil {
+			return nil, fmt.Errorf("compute hash: %w", err)
+		}
+		fieldGroup2hash[fg.Name] = h
+	}
+
+	// Compute the total hash
+	hash := fnv.New32a()
+	for _, fgName := range keyFields.Hash.FieldGroups {
```

Comment:
Can you explain somewhere why we need different FieldGroups in KeyFields and in Hash? What is the purpose of these additional FieldGroups?

Reply:
One is of type []FieldGroup and one is of type []string. Using the same name for these 2 different fields is a bit confusing. I assume that the []string is derived from Name field of the []FieldGroup.

Reply:
You are correct.
`KeyFields.FieldGroups` is where you define the field groups
and
`KeyFields.Hash.FieldGroups` is where you reference the defined field groups (by name). I'll rename it and add some comments to make it a bit clearer.


Please take a look at these configuration examples:
https://gist.github.com/ronensc/0136cd1c3a761e2d6067679362fa72c6
https://gist.github.com/ronensc/c6d0a07b0503bbd7a36bc4ef0b6b5b29